options.md
# Options in Soft Matrix

Soft Matrix has a few options for configuring the generated wave file and how it processes sound.

## Output Options

**-matrix**: Chooses the matrix to use. Available matrixes are:

- **default**: The default matrix, used when the "-matrix" option is omitted. Sounds that are out-of-phase are panned to the rear. Sounds that are in phase are panned to the front. A good "all-round" matrix for recordings with significant out-of-phase material.
- **horseshoe**: Intended for recordings that are mostly panned between the two speakers, without much out-of-phase material. Widening is applied, and sounds that are in the extreme right and left are panned to the rear. Out-of-phase material is also panned to the rear.
- **dolby**: Adheres to the Dolby Stereo matrix, also known as LtRt, Dolby Surround, and Dolby Pro-Logic. Dolby Stereo was used on analog soundtracks for theatrical movies starting in the late 1970s, and was also used in analog television. When recordings encoded in Dolby Digital are downmixed to stereo, they are often matrixed using Dolby Stereo. (See <https://en.wikipedia.org/wiki/Dolby_Stereo#The_Dolby_Stereo_Matrix> for more information.)
- **qs**: Adheres to the "qs" matrix. Very similar to the "default" matrix, except that some minor widening is present. (See <https://en.wikipedia.org/wiki/QS_Regular_Matrix> for more information.)
- **rm**: Synonym for "qs". This option exists because it was common to mislabel qs-encoded recordings as rm.
- **sq**: EXPERIMENTAL! Adheres to the "sq" matrix. Although this matrix had a lot of commercial releases in the late 1970s, its technical limitations held it back from widespread adoption. Due to SQ's flaws, this option should only be used on material explicitly encoded for SQ. (See <https://en.wikipedia.org/wiki/Stereo_Quadraphonic>). (Note that sq support is experimental. This approach closely inspects phase and amplitude, but doesn't decode very well.)
- **sqexperimental**: An experimental decoder for sq that preserves in-phase front tones very well, and then uses a "by the book" dematrixer when
tones aren't in phase. This also works poorly. It may be removed in a future release of Soft Matrix.

**-channels**: The channel layout in the output file

- **4**: Four-channel layout; quadraphonic. Includes front right and left; and rear front and left.
- **5**: Five-channel layout. Includes front right, center, and left; and rear front and left.
- **5.1**: Five-point-one channel layout. Includes front right, center, and left; rear front and left; and a subwoofer channel.

**-minimum**: The minimum amplitude to steer front-to-back. Defaults to 0.01. On very clean signals, it may be useful to use a lower
threshold, like 0.0001. (This is needed because sounds that are isolated into the right front or right left speaker may be mis-steered due to the phase of noise in the adjacent source channel.)

**-loud**: Does not lower the amplitude when generating a center or LFE channel. [Because a center or LFE channel is based off of mixing the right and left channels, the overall amplitude is lowered in order to avoid clipping.](<Documentation/The loud flag.md>) This setting is useful when upmixing source material that is quiet, or otherwise mixed in a way to prevent clipping when upmixed. (Upmixing to 4.0 defaults to loud). (Not valid for 4.0.)

**-quiet**: Lowers the amplitude. (Default behavior for 4.1, 5.0, and 5.1.)

## Performance Options

**-low**: Specifies the lowest frequency calculated in the matrix. (Defaults to 20 hz.) Steering lower frequencies will make Soft Matrix run very slowly. If this is set too high, it may impede calculating the subwoofer or steering audible frequencies. (Very low frequencies require a much larger window for Fourier transforms. Larger windows take significantly longer to calculate.)

**-threads**: The number of threads to run. Defaults to [available_parallelism()](https://doc.rust-lang.org/stable/std/thread/fn.available_parallelism.html). This option is useful because available_parallelism() may return a number lower than the number of cores present in the CPU. Setting this higher than the number of cores in your CPU is not advised. This is a useful option if soft_matrix makes your computer run slowly.

**-keepawake**: Controls if soft_matrix keeps the computer awake. When true, the computer is prevented from sleeping while soft_matrix is running. When false, the computer can sleep while idle. Defaults to true.

## Examples

### Upmix a wave file using all defaults

    soft_matrix "stereo.wav" "surround.wav"

This will upmix stereo.wav, the input file, to a 5.1 wav file named surround.wav, using the default matrix.

### Use the RM matrix

    soft_matrix "stereo.wav" "surround.wav" -matrix rm

This will upmix stereo.wav, the input file, to a 5.1 wav file named surround.wav, using the RM matrix.

### Only run a single thread

    soft_matrix "stereo.wav" "surround.wav" -threads 1

This will upmix stereo.wav, the input file, using only a single thread. This option is useful if soft_matrix is making your computer run slowly.

### Run a fast encode without an LFE channel

    soft_matrix "stereo.wav" "preview.wav" -low 60 -channels 5

This only steers frequencies above 60 hz. Useful for a quick preview of upmixing.

### Allow the computer to sleep while upmixing

    soft_matrix "stereo.wav" "surround.wav" -keepawake false

This will allow the computer to sleep while upmixing. (Default behavior is that the computer will not sleep while running.)


Releases.md
# Release History

## 1.0.5

Fixed compiler warning

## 1.0.4

Added URL at startup
Added articles:

- [How is Stereo Upmixed to Surround Sound](https://github.com/GWBasic/soft_matrix/blob/main/Documentation/How%20is%20Stereo%20Upmixed%20to%20Surround%20Sound.md)
- [Samples](https://github.com/GWBasic/soft_matrix/blob/main/Documentation/Samples.md)
- [The loud flag](https://github.com/GWBasic/soft_matrix/blob/main/Documentation/The%20loud%20flag.md)

## 1.0.3

Documentation

## 1.0.2

Documentation

## 1.0.1

[Display version number at startup](https://github.com/GWBasic/soft_matrix/issues/92)

## 1.0.0

Initial Release


src/options.rs
use std::env;
use std::path::Path;

use wave_stream::wave_header::Channels;

use crate::{
    matrix::{DefaultMatrix, Matrix, SQMatrix, SQMatrixExperimental},
    panner_and_writer,
};

pub struct Options {
    pub source_wav_path: Box<Path>,
    pub target_wav_path: Box<Path>,
    pub num_threads: Option<usize>,
    pub transform_mono: bool,
    pub channels: Channels,
    pub low_frequency: f32,
    pub minimum_steered_amplitude: f32,
    pub keep_awake: bool,
    pub loud: bool,

    // Performs additional adjustments according to the specific chosen matrix
    // SQ, QS, RM, ect
    pub matrix: Box<dyn Matrix>,
}

pub enum ChannelLayout {
    Four,
    Five,
    FiveOne,
}

pub enum MatrixFormat {
    Default,
    QS,
    HorseShoe,
    DolbyStereo,
    SQ,
    SQExperimental,
}

impl Options {
    pub fn parse() -> Option<Options> {
        let args: Vec<String> = env::args().collect();

        if args.len() < 3 {
            println!("Usage: soft_matrix [source] [destination]");
            return None;
        }

        let mut args_iter = args.into_iter();

        // ignore the executable name
        let _ = args_iter.next().unwrap();

        let source_wav_path = args_iter.next().unwrap();
        let source_wav_path = Path::new(source_wav_path.as_str());

        let target_wav_path = args_iter.next().unwrap();
        let target_wav_path = Path::new(target_wav_path.as_str());

        let mut num_threads = None;

        let mut channel_layout = ChannelLayout::FiveOne;
        let mut matrix_format = MatrixFormat::Default;
        let mut low_frequency = 20.0f32;

        let mut minimum_steered_amplitude = 0.01;

        let mut keep_awake = true;

        let mut loud: Option<bool> = None;

        // Iterate through the options
        // -channels
        // 4 or 5 or 5.1

        loop {
            match args_iter.next() {
                Some(flag) => {
                    // Parse a flag
                    if flag.eq("-channels") {
                        match args_iter.next() {
                            Some(channels_string) => {
                                if channels_string.eq("4") {
                                    channel_layout = ChannelLayout::Four
                                } else if channels_string.eq("5") {
                                    channel_layout = ChannelLayout::Five
                                } else if channels_string.eq("5.1") {
                                    channel_layout = ChannelLayout::FiveOne
                                } else {
                                    println!("Unknown channel configuration: {}", channels_string);
                                    return None;
                                }
                            }
                            None => {
                                println!("Channels unspecified");
                                return None;
                            }
                        }
                    } else if flag.eq("-matrix") {
                        match args_iter.next() {
                            Some(matrix_format_string) => {
                                if matrix_format_string.eq("default") {
                                    matrix_format = MatrixFormat::Default
                                } else if matrix_format_string.eq("qs") {
                                    matrix_format = MatrixFormat::QS
                                } else if matrix_format_string.eq("rm") {
                                    matrix_format = MatrixFormat::QS
                                } else if matrix_format_string.eq("horseshoe") {
                                    matrix_format = MatrixFormat::HorseShoe
                                } else if matrix_format_string.eq("dolby") {
                                    matrix_format = MatrixFormat::DolbyStereo
                                } else if matrix_format_string.eq("sq") {
                                    matrix_format = MatrixFormat::SQ
                                } else if matrix_format_string.eq("sqexperimental") {
                                    matrix_format = MatrixFormat::SQExperimental
                                } else {
                                    println!("Unknown matrix format: {}", matrix_format_string);
                                    return None;
                                }
                            }
                            None => {
                                println!("Matrix unspecified");
                                return None;
                            }
                        }
                    } else if flag.eq("-low") {
                        match args_iter.next() {
                            Some(low_frequency_string) => {
                                match low_frequency_string.parse::<f32>() {
                                    Ok(low_frequency_arg) => {
                                        if low_frequency_arg < 1.0 {
                                            println!(
                                                "Lowest frequency must >= 1: {}",
                                                low_frequency_arg
                                            );
                                            return None;
                                        }

                                        low_frequency = low_frequency_arg
                                    }
                                    Err(_) => {
                                        println!(
                                            "Lowest frequency must be an integer: {}",
                                            low_frequency_string
                                        );
                                        return None;
                                    }
                                }
                            }
                            None => {
                                println!("Lowest frequency unspecified");
                                return None;
                            }
                        }
                    } else if flag.eq("-threads") {
                        match args_iter.next() {
                            Some(num_threads_string) => match num_threads_string.parse::<usize>() {
                                Ok(num_threads_value) => num_threads = Some(num_threads_value),
                                Err(_) => {
                                    println!(
                                        "Can not parse the number of threads: {}",
                                        num_threads_string
                                    );
                                    return None;
                                }
                            },
                            None => {
                                println!("Number of threads unspecified");
                                return None;
                            }
                        }
                    } else if flag.eq("-minimum") {
                        match args_iter.next() {
                            Some(minimum_steered_amplitude_string) => {
                                match minimum_steered_amplitude_string.parse::<f32>() {
                                    Ok(minimum_steered_amplitude_value) => {
                                        minimum_steered_amplitude = minimum_steered_amplitude_value
                                    }
                                    Err(_) => {
                                        println!(
                                            "Can not parse the minimum amplitude: {}",
                                            minimum_steered_amplitude_string
                                        );
                                        return None;
                                    }
                                }
                            }
                            None => {
                                println!("Minimum amplitude unspecified");
                                return None;
                            }
                        }
                    } else if flag.eq("-keepawake") {
                        match args_iter.next() {
                            Some(keep_awake_string) => match keep_awake_string.parse::<bool>() {
                                Ok(keep_awake_value) => keep_awake = keep_awake_value,
                                Err(_) => {
                                    println!(
                                        "Can not parse the keep awake value: {}",
                                        keep_awake_string
                                    );
                                    return None;
                                }
                            },
                            None => {
                                println!("Keepawake value unspecified");
                                return None;
                            }
                        }
                    } else if flag.eq("-loud") {
                        loud = Some(true);
                    } else if flag.eq("-quiet") {
                        loud = Some(false);
                    } else {
                        println!("Unknown flag: {}", flag);
                        return None;
                    }
                }
                None => {
                    // No more flags left, interpret the options and return them
                    let transform_mono: bool;
                    let channels: Channels;

                    match channel_layout {
                        ChannelLayout::Four => {
                            transform_mono = false;
                            channels = Channels::new()
                                .front_left()
                                .front_right()
                                .back_left()
                                .back_right();
                        }
                        ChannelLayout::Five => {
                            transform_mono = true;
                            channels = Channels::new()
                                .front_left()
                                .front_right()
                                .front_center()
                                .back_left()
                                .back_right();
                        }
                        ChannelLayout::FiveOne => {
                            transform_mono = true;
                            channels = Channels::new()
                                .front_left()
                                .front_right()
                                .front_center()
                                .low_frequency()
                                .back_left()
                                .back_right();
                        }
                    }

                    let matrix: Box<dyn Matrix> = match matrix_format {
                        MatrixFormat::Default => Box::new(DefaultMatrix::new()),
                        MatrixFormat::QS => Box::new(DefaultMatrix::qs()),
                        MatrixFormat::HorseShoe => Box::new(DefaultMatrix::horseshoe()),
                        MatrixFormat::DolbyStereo => Box::new(DefaultMatrix::dolby_stereo()),
                        MatrixFormat::SQ => Box::new(SQMatrix::sq()),
                        MatrixFormat::SQExperimental => Box::new(SQMatrixExperimental::sq()),
                    };

                    if (low_frequency as f32) > panner_and_writer::LFE_START
                        && channels.low_frequency
                    {
                        println!(
                            "LFE channel not supported when the lowest frequency to steer ({}hz) is greater than {}hz",
                            low_frequency,
                            panner_and_writer::LFE_START);
                        return None;
                    }

                    let loud = if transform_mono {
                        loud.unwrap_or(false)
                    } else {
                        if loud.is_some() {
                            println!("-loud and -quiet only work when upmixing with an LFE or a center channel");
                            return None;
                        }

                        true
                    };

                    return Some(Options {
                        source_wav_path: source_wav_path.into(),
                        target_wav_path: target_wav_path.into(),
                        num_threads,
                        transform_mono,
                        channels,
                        matrix,
                        low_frequency,
                        minimum_steered_amplitude,
                        keep_awake,
                        loud,
                    });
                }
            }
        }
    }
}
